### ISSUE TITLE
Is performance important enough for the MetricDescriptor Type to encode it as a unit64?

### ISSUE THREAD

**Issue by bogdandrutu** [open-telemetry/opentelemetry-proto#189]
Currently there is a TODO in the metrics proto that needs a decision:

> // TODO: Determine if encoding all possible values in a uint64 bitset improves performance significantly and propose that if that is the case.

**Comment by jmacd** [comment]
We have discussed that the performance implications of this are more important in the collector. Clients can cache descriptor objects and/or use alternate encoders fairly easily.

**Comment by tigrannajaryan** [comment]
It is very likely that there will be significant performance gains for single datapoint metrics. Benchmarking using https://github.com/tigrannajaryan/exp-otelproto should be easy.

**Comment by tigrannajaryan** [comment]
>Clients can cache descriptor objects and/or use alternate encoders fairly easily.

True, but I suspect most clients won't bother, so would be good not to rely on this. Using bitmasks is trivial in most (?) languages, so I don't think usability-wise it is any worse than `oneof` with a bunch of messages (in fact my preference would be a single bitfield with masks regardless of performance).

**Comment by jmacd** [comment]
(In #168 I wrote a program to generate the bits of all valid combinations...)

**Comment by bogdandrutu** [comment]
@tigrannajaryan are you most worried about bytes on the wire or number of allocations?

**Comment by bogdandrutu** [comment]
As an example the current Gauge does not have temporality and we tried to add a new temporality just for that, then found hard to reason about all the corner cases

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan are you most worried about bytes on the wire or number of allocations?

The later.

> The biggest problem with PR #168 was that it makes things harder to use since not every possible combination in a bitmap is valid.

Can we provide valid combinations of bitmasks as enums/consts and clearly document that other combinations are not valid?

**Comment by bogdandrutu** [comment]
> The later.

Let me give a try to use gogo proto to avoid allocations. If that works we should not worry and complicate things.
